spring-integration-core/src/main/java/org/springframework/integration/router/ErrorMessageExceptionTypeRouter.java

```diff
@@ -38,15 +46,28 @@
 		Object payload = message.getPayload();
 		if (payload instanceof Throwable) {
 			Throwable cause = (Throwable) payload;
+			Set<String> mapping = getChannelMappings().keySet();
+			Map<String, Class<?>> classesCache = new HashMap<String, Class<?>>();
 			while (cause != null) {
-				String causeName = cause.getClass().getName();
-				if (this.getChannelMappings().keySet().contains(causeName)) {
-					mostSpecificCause = causeName;
+				for (String className : mapping) {
+					Class<?> exceptionClass = classesCache.get(className);
+					if (exceptionClass == null) {
+						try {
+							exceptionClass = ClassUtils.forName(className, getApplicationContext().getClassLoader());
+							classesCache.put(className, exceptionClass);
+						}
+						catch (ClassNotFoundException e) {
+							throw new IllegalStateException(e);
```

Comment:
I think we should detect bad keys during initialization rather than runtime, store the cache as a field, and simply log rather than throwing an exception. The runtime exception can be controlled by `resolutionRequired`.

We certainly shouldn't build the cache on every call, but it needs to be cleared/rebuilt whenever the mappings change.